### DISCUSSION TITLE
Predictable & sane debugging in multithreaded apps (i.e. those using go routines or unsafe{} stuff)

### DISCUSSION THREAD

**Discussion by dumblob** [vlang/v#11707] [Ideas]
I'm still on my way through all the materials from @cristian-ilies-vasile and after finally discovering where the implementation behind [Bounding Data Races in Space and Time](https://kcsrk.info/papers/pldi18-memory.pdf ) (from https://kcsrk.info/publications.html ) lies, I'd like to **propose the same measures for *V***'s debug build target (definitely not for production builds though).

Basically the paper boils down to the fact, that if there is any data race between two or more threads, then it's extremely hard to debug if the C/C++ "relaxed" memory model has been used. In other words, any debugging technique (including the good old "*printf() everywhere*") will produce pretty **random-looking** results. To avoid this debugging nightmare, the memory model has to be slightly less relaxed.

To our advantage, authors of the paper implemented this less relaxed model on top of the C/C++ memory model (they did this implementation for OCaml, but that's not important here), so we can basically copy & paste it to *V*. The PR is here: https://github.com/ocaml-multicore/ocaml-multicore/pull/225 .

@cristian-ilies-vasile @UweKrueger @helto4real @medvednikov @spytheman might be of interest for you

**Comment by cristian-ilies-vasile** [comment]
@dumblob 
Thanks a lot!

**Comment by cristian-ilies-vasile** [comment]
@dumblob 
When you have time, look at https://divine.fi.muni.cz/index.html
_DIVINE is a modern, explicit-state model checker. Based on the LLVM toolchain, it can verify programs written in multiple real-world programming languages, including C and C++. The verification core is built on a foundation of high­-per­for­mance algorithms and data structures, scaling all the way from a laptop to a high-end cluster._

**Comment by dumblob** [comment]
@cristian-ilies-vasile didn't know about DIVINE. According to their description it's a mature piece of verification software. And what is striking, it actually **fully support** parallelism (os threads) for x86-TSO memory model (other memory models are ignored - e.g. those on ARM etc.). I think this could be very useful to run before each release of *V* to find races in *V* compiler itself. But I think it'd be impractical to run more often because it's quite slow.

Good catch anyway!

**Comment by dumblob** [comment]
@medvednikov why was this moved to discussions? There is a clear actionable item. Namely to implement the less-relaxed semantics of the memory model for debug builds.
